tests/bsim/bluetooth/host/adv/long_ad/src/advertiser.c

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdbool.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/logging/log.h>
+#include <sys/types.h>
+
+#include "ad.h"
+#include "babblekit/testcase.h"
+
+#define MAX_AD 10
+
+LOG_MODULE_REGISTER(advertiser, LOG_LEVEL_ERR);
+
+static void create_adv(struct bt_le_ext_adv **adv)
+{
+	int err;
+	struct bt_le_adv_param params;
+
+	memset(&params, 0, sizeof(struct bt_le_adv_param));
+
+	params.options |= BT_LE_ADV_OPT_EXT_ADV;
+
+	params.id = BT_ID_DEFAULT;
+	params.sid = 0;
+	params.interval_min = BT_GAP_ADV_SLOW_INT_MIN;
+	params.interval_max = BT_GAP_ADV_SLOW_INT_MAX;
+
+	err = bt_le_ext_adv_create(&params, NULL, adv);
+	if (err) {
+		TEST_FAIL("Failed to create advertiser (%d)\n", err);
+	}
+}
+
+static void start_adv(struct bt_le_ext_adv *adv)
+{
+	int err;
+	int32_t timeout = 0;
+	uint8_t num_events = 0;
+
+	struct bt_le_ext_adv_start_param start_params;
+
+	start_params.timeout = timeout;
+	start_params.num_events = num_events;
+
+	err = bt_le_ext_adv_start(adv, &start_params);
+	if (err) {
+		TEST_FAIL("Failed to start advertiser (%d)\n", err);
+	}
+}
+
+static size_t ad_deserialize(const uint8_t *input, size_t input_size, struct bt_data output[])
+{
+	uint8_t type;
+	uint8_t data_len;
+	const uint8_t *data;
+
+	size_t idx = 0;
+	size_t ad_len = 0;
+
+	while (idx != input_size) {
+		data_len = input[idx] - 1;
+		type = input[idx + 1];
+		data = &input[idx + 2];
+
+		if (data_len + 2 > input_size - idx) {
+			TEST_FAIL("malformed advertising data, expected %d bytes of data but got "
+				  "only %d "
+				  "bytes",
+				  data_len + 1, input_size - idx);
+			return -1;
+		}
+
+		output[ad_len].data = data;
+		output[ad_len].type = type;
+		output[ad_len].data_len = data_len;
+
+		ad_len += 1;
+		idx += data_len + 2;
+	}
+
+	return ad_len;
+}
+
+static int set_ad_data(struct bt_le_ext_adv *adv, const uint8_t *serialized_ad,
+			size_t serialized_ad_size)
+{
+	int err;
+	size_t ad_len = 0;
+
+	struct bt_data ad[MAX_AD];
```

Comment:
nit: maybe it makes more sense to move that next to the define. or just remove the define altogether